#### Fase 2: Transformación de los datos

### 1. Importación de librerías

In [248]:
import numpy as np
import pandas as pd
from sklearn.impute import SimpleImputer
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer
from sklearn.impute import KNNImputer
import re
from word2number import w2n
pd.set_option('display.max_columns', None)

### 2. Importación del csv y exclusión de las columnas innecesarias

In [249]:
#Importamos csv
url=('https://raw.githubusercontent.com/s-armeni/proyecto-da-promo-H-modulo-3-team-3-DataMinds/main/HR%20RAW%20DATA.csv')
df_raw=pd.read_csv(url)
df_raw.head()

Unnamed: 0 Age Attrition BusinessTravel DailyRate  \
0           0  51        No            NaN    684,0$   
1           1  52        No            NaN    699,0$   
2           2  42        No  travel_rarely    532,0$   
3           3  47        No  travel_rarely    359,0$   
4           4  46        No            NaN   1319,0$   

                 Department  DistanceFromHome  Education    EducationField  \
0                       NaN                 6          3               NaN   
1                       NaN                 1          4     Life Sciences   
2   Research & Development                  4          2  Technical Degree   
3                       NaN                 2          4           Medical   
4                       NaN                 3          3  Technical Degree   

   employeecount employeenumber  EnvironmentSatisfaction  Gender HourlyRate  \
0              1          162,0                        1       0         51   
1              1          259,0                        3       0         65   
2              1          319,0                        3       0         58   
3              1            NaN                        1       1         82   
4              1            NaN                        1       1         45   

   JobInvolvement  JobLevel              JobRole  JobSatisfaction  \
0               3         5   resEArch DIREcToR                 3   
1               2         5             ManAGeR                 3   
2               3         5             ManaGER                 4   
3               3         4   ReseArCH DIrECtOr                 3   
4               4         4     sAleS EXECUtIve                 1   

  MaritalStatus MonthlyIncome  MonthlyRate  NUMCOMPANIESWORKED Over18  \
0           NaN       19537,0         6462                   7      Y   
1           NaN       19999,0         5678                   0    NaN   
2       Married       19232,0         4933                   1    NaN   
3       Married       17169,0        26703                   3      Y   
4      Divorced           NaN         7739                   2      Y   

  OverTime  PercentSalaryHike PerformanceRating  RelationshipSatisfaction  \
0       No                 13               3,0                         3   
1      NaN                 14               3,0                         1   
2       No                 11               3,0                         4   
3      NaN                 19               3,0                         2   
4       No                 12               3,0                         4   

  StandardHours  StockOptionLevel TOTALWORKINGYEARS  TrainingTimesLastYear  \
0           NaN                 0               NaN                      5   
1           NaN                 1              34,0                      5   
2           NaN                 0              22,0                      3   
3           NaN                 2               NaN                      2   
4           NaN                 1               NaN                      5   

  WORKLIFEBALANCE  YearsAtCompany YearsInCurrentRole  YearsSinceLastPromotion  \
0             3,0              20                NaN                       15   
1             3,0              33                NaN                       11   
2             NaN              22                NaN                       11   
3             NaN              20                NaN                        5   
4             3,0              19                NaN                        2   

   YEARSWITHCURRMANAGER SameAsMonthlyIncome  DateBirth       Salary  \
0                    15             19537,0       1972  1000000000$   
1                     9             19999,0       1971  1000000000$   
2                    15             19232,0       1981  1000000000$   
3                     6             17169,0       1976  1000000000$   
4                     8                 NaN       1977  1000000000$   

                      

In [250]:
#Creamos nueva variable para guardar la selección del rango de columnas que queremos.
#Este será el csv que utlizaremos para la limpieza
df_data=df_raw.iloc[:, 1:]
df_data.head()

Age Attrition BusinessTravel DailyRate                Department  \
0  51        No            NaN    684,0$                       NaN   
1  52        No            NaN    699,0$                       NaN   
2  42        No  travel_rarely    532,0$   Research & Development    
3  47        No  travel_rarely    359,0$                       NaN   
4  46        No            NaN   1319,0$                       NaN   

   DistanceFromHome  Education    EducationField  employeecount  \
0                 6          3               NaN              1   
1                 1          4     Life Sciences              1   
2                 4          2  Technical Degree              1   
3                 2          4           Medical              1   
4                 3          3  Technical Degree              1   

  employeenumber  EnvironmentSatisfaction  Gender HourlyRate  JobInvolvement  \
0          162,0                        1       0         51               3   
1          259,0                        3       0         65               2   
2          319,0                        3       0         58               3   
3            NaN                        1       1         82               3   
4            NaN                        1       1         45               4   

   JobLevel              JobRole  JobSatisfaction MaritalStatus MonthlyIncome  \
0         5   resEArch DIREcToR                 3           NaN       19537,0   
1         5             ManAGeR                 3           NaN       19999,0   
2         5             ManaGER                 4       Married       19232,0   
3         4   ReseArCH DIrECtOr                 3       Married       17169,0   
4         4     sAleS EXECUtIve                 1      Divorced           NaN   

   MonthlyRate  NUMCOMPANIESWORKED Over18 OverTime  PercentSalaryHike  \
0         6462                   7      Y       No                 13   
1         5678                   0    NaN      NaN                 14   
2         4933                   1    NaN       No                 11   
3        26703                   3      Y      NaN                 19   
4         7739                   2      Y       No                 12   

  PerformanceRating  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0               3,0                         3           NaN                 0   
1               3,0                         1           NaN                 1   
2               3,0                         4           NaN                 0   
3               3,0                         2           NaN                 2   
4               3,0                         4           NaN                 1   

  TOTALWORKINGYEARS  TrainingTimesLastYear WORKLIFEBALANCE  YearsAtCompany  \
0               NaN                      5             3,0              20   
1              34,0                      5             3,0              33   
2              22,0                      3             NaN              22   
3               NaN                      2             NaN              20   
4               NaN                      5             3,0              19   

  YearsInCurrentRole  YearsSinceLastPromotion  YEARSWITHCURRMANAGER  \
0                NaN                       15                    15   
1                NaN                       11                     9   
2                NaN                       11                    15   
3                NaN                        5                     6   
4                NaN                        2                     8   

  SameAsMonthlyIncome  DateBirth       Salary  \
0             19537,0       1972  1000000000$   
1             19999,0       1971  1000000000$   
2             19232,0       1981  1000000000$   
3             17169,0       1976  1000000000$   
4                 NaN       1977  1000000000$   

                        RoleDepartament  NUMBERCHILDREN RemoteWork  
0                       

In [251]:
#Exportamos el csv para tener guardada la versión con la que vamos a trabajar.
df_data.to_csv('hr_data.csv', index=False)

In [252]:
#Visualizamos el número total de filas y de columnas
print(f'Cantidad de filas: {df_data.shape[0]}\nCantidad de columnas: {df_data.shape[1]}')

Cantidad de filas: 1614
Cantidad de columnas: 41


### 3. Estandarización de los nombres de las columnas

3.1. Lo primero que queremos hacer es separar las palabras mediante '_'

In [253]:
#Vemos los actuales nombres de las columnas
df_data.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'employeecount',
       'employeenumber', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NUMCOMPANIESWORKED',
       'Over18', 'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TOTALWORKINGYEARS', 'TrainingTimesLastYear', 'WORKLIFEBALANCE',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YEARSWITHCURRMANAGER', 'SameAsMonthlyIncome', 'DateBirth', 'Salary',
       'RoleDepartament', 'NUMBERCHILDREN', 'RemoteWork'],
      dtype='object')

In [254]:
#Variable con los nombres de columna originales
old_columns=df_data.columns.tolist()

#Creamos variable con las palabras separadas mediante '_'
new_columns=['Age', 'Attrition', 'Business_Travel', 'Daily_Rate', 'Department',
       'Distance_From_Home', 'Education', 'Education_Field', 'employee_count',
       'employee_number', 'Environment_Satisfaction', 'Gender', 'Hourly_Rate',
       'Job_Involvement', 'Job_Level', 'Job_Role', 'Job_Satisfaction',
       'Marital_Status', 'Monthly_Income', 'Monthly_Rate', 'NUM_COMPANIES_WORKED',
       'Over_18', 'Over_Time', 'Percent_Salary_Hike', 'Performance_Rating',
       'Relationship_Satisfaction', 'Standard_Hours', 'Stock_Option_Level',
       'TOTAL_WORKING_YEARS', 'Training_Times_Last_Year', 'WORK_LIFE_BALANCE',
       'Years_At_Company', 'Years_In_Current_Role', 'Years_Since_Last_Promotion',
       'YEARS_WITH_CURR_MANAGER', 'Same_As_Monthly_Income', 'Date_Birth', 'Salary',
       'Role_Departament', 'NUMBER_CHILDREN', 'Remote_Work']
#Hacemos un if para asegurarnos que no se queda ninguna columna sin nombre
if len(new_columns)==len(df_data.columns):
    df_data.columns=new_columns
else:
    raise ValueError('La lista new_columns debe tener la misma longitud que número de columnas tiene el DataFrame.')

#Visualizamos cabezara para comprobar que se han hecho los cambios
df_data.head(1)

Age Attrition Business_Travel Daily_Rate Department  Distance_From_Home  \
0  51        No             NaN     684,0$        NaN                   6   

   Education Education_Field  employee_count employee_number  \
0          3             NaN               1           162,0   

   Environment_Satisfaction  Gender Hourly_Rate  Job_Involvement  Job_Level  \
0                         1       0          51                3          5   

              Job_Role  Job_Satisfaction Marital_Status Monthly_Income  \
0   resEArch DIREcToR                  3            NaN        19537,0   

   Monthly_Rate  NUM_COMPANIES_WORKED Over_18 Over_Time  Percent_Salary_Hike  \
0          6462                     7       Y        No                   13   

  Performance_Rating  Relationship_Satisfaction Standard_Hours  \
0                3,0                          3            NaN   

   Stock_Option_Level TOTAL_WORKING_YEARS  Training_Times_Last_Year  \
0                   0                 NaN                         5   

  WORK_LIFE_BALANCE  Years_At_Company Years_In_Current_Role  \
0               3,0                20                   NaN   

   Years_Since_Last_Promotion  YEARS_WITH_CURR_MANAGER Same_As_Monthly_Income  \
0                          15                       15                19537,0   

   Date_Birth       Salary Role_Departament  NUMBER_CHILDREN Remote_Work  
0        1972  1000000000$              NaN              NaN         Yes

3.2. Unificar en capitalize.

In [255]:
df_data.columns=df_data.columns.str.capitalize()

### 4. Estandarización de los datos en las filas

In [256]:
#Vemos el tipo de dato de cada columna para identificar aquellos que tendremos que modificar
df_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1614 entries, 0 to 1613
Data columns (total 41 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Age                         1614 non-null   object 
 1   Attrition                   1614 non-null   object 
 2   Business_travel             842 non-null    object 
 3   Daily_rate                  1614 non-null   object 
 4   Department                  302 non-null    object 
 5   Distance_from_home          1614 non-null   int64  
 6   Education                   1614 non-null   int64  
 7   Education_field             869 non-null    object 
 8   Employee_count              1614 non-null   int64  
 9   Employee_number             1183 non-null   object 
 10  Environment_satisfaction    1614 non-null   int64  
 11  Gender                      1614 non-null   int64  
 12  Hourly_rate                 1614 non-null   object 
 13  Job_involvement             1614 

4.1 Estandarización de la columna 'Age'.

In [257]:
#Función para pasar los números de object a int, y los números formato 'forty-seven' a int
def age_to_int(num):
    try:
        return int(num)
    except ValueError:
        try:
            return w2n.word_to_num(num)
        except ValueError:
            return np.nan
df_data['Age']=df_data['Age'].apply(age_to_int)

4.2 Estandarización de la columna 'Distance_from_home'.

In [258]:
#Eliminamos los '-' de 'Distance_from_home'
df_data.loc[df_data['Distance_from_home']<0, 'Distance_from_home']*=-1

4.3 Estandarización de la columna 'Marital_status'.

In [259]:
#Corregimos errores ortográficos en la columna 'Marital_status'
df_data['Marital_status']=df_data['Marital_status'].replace('Marreid','married')

4.4 Estandarización de la columna'Environment_satisfaction'.

In [260]:
#Definimos una funcion:
def extrapolate(valor):
    if 4>=valor:
        return valor
    elif 4<valor<=10:
        return 1
    elif 10<valor<=20:
        return 1
    elif 20<valor<=30:
        return 2
    elif 30<valor<=40:
        return 3
    elif valor>40:
        return 4

#Aplicamos la funcion anterior a la columna 'Environment_satisfaction'
df_data['Environment_satisfaction']=df_data['Environment_satisfaction'].apply(extrapolate)

4.5 Eliminamos % de la columna 'Daily_rate'.

In [261]:
#Función para eliminar los $
def remove_dollar(df, columns):
    for i in columns:
        df[i]=df[i].str.replace('$', '', regex=False)
    return df

#En esta lista están todas las columnas a las que tenemos que aplicar la función
list_dollar=['Daily_rate', 'Salary']

#Llamamos a la función para modificar la lista
df_data=remove_dollar(df_data, list_dollar)

4.6 Sustituimos 'Not Available' por 'NaN' en la columna 'Hourly_rate'

In [262]:
#Para poder operar con la media y transformalos en float, es necesario pasar el 'not available' a valores nulo nan:
def change_to_null (string):
    if string == 'Not Available':
        return np.nan
    else:
        return string

df_data['Hourly_rate']=df_data['Hourly_rate'].apply(change_to_null)

4.7 Agrupamos las columnas por tipo de limpieza/correción que necesitan.

In [263]:
#Lista de columnas integer a object hombre/mujer
list_male_female=['Gender']

#Lista de columnas yes/no
list_yes_no=['Attrition', 'Over_18', 'Over_time', 'Remote_work']

#Lista de columnas integer a float
list_int_float=['Distance_from_home', 'Employee_count', 'Monthly_rate', 'Percent_salary_hike']

#Lista de columnas object a float
list_object_float=['Daily_rate', 'Hourly_rate', 'Monthly_income', 'Standard_hours', 'Same_as_monthly_income', 'Salary']

#Lista de columnas object a integer
list_object_integer=['Employee_number', 'Performance_rating', 'Total_working_years', 'Work_life_balance', 'Years_in_current_role']

En las siguientes columnas no necesitamos hacer limpieza:
- Business_travel
- Department
- Education
- Education_field
- Job_involvement
- Job_level
- Job_role
- Job_satisfaction
- Num_companies_worked
- Relationship_satisfaction
- Stock_option_level
- Training_times_last_year
- Years_at_company
- Years_since_last_promotion
- Years_with_curr_manager
- Date_birth
- Role_departament
- Number_children

4.8 Eliminamos posibles espacios en blanco al principio y final de las celdas.

In [264]:
#Función para eliminar los espacios en blanco al principio y al final de los strings
def strip_strings(cell):
    if isinstance(cell, str): #si es un string retorna sin espacios
        return cell.strip()
    return cell

#Aplicamos la función a todas las celdas del DataFrame
df_data=df_data.applymap(strip_strings) #.applymap() porque lo aplico a cada elemento del df

4.9 Estandarizamos la columna de género.

In [265]:
#Función para reemplazar 0 por M y 1 por F
def male_female(df, column):
    df[column]=df[column].replace({0:'Male', 1:'Female'})
    return df

#Aplicamos la función a la columna gender
df_data=male_female(df_data, 'Gender')

4.10 Estandarizamos las columnas de 'Yes/No'.

In [266]:
#Función para unificar yes/no
def unify_yes_no(df, columns):
    yes_no_map={'yes':'Yes', 'YES':'Yes', 'TRUE':'Yes', 'true':'Yes', '1':'Yes', 'Y':'Yes', 'Yes':'Yes',
        'no':'No', 'No':'No', 'NO':'No', 'FALSE':'No', 'false':'No', '0':'No', 'False':'No'}
    
    for i in columns:
        df[i]=df[i].astype(str).map(yes_no_map)
    return df

#Llamamos a la función para modificar la lista
df_data=unify_yes_no(df=df_data, columns=list_yes_no)

4.11 Estandarizamos columnas integer a float.

In [267]:
#Función para convertir int to float
def int_to_float(df, columns):
    for i in columns:
        df[i]=df[i].astype(float).round(2)
    return df
    
#Llamamos a la función para modificar la lista
df_data=int_to_float(df_data, list_int_float)

4.12 Estandarizamos columnas object a float.

In [268]:
#Función para convertir object a float
def object_to_float(df, columns):
    for i in columns:
        try:
            df[i]=df[i].str.replace(',', '.').astype(float).round(2)
        except:
            df[i]=df[i].astype(float).round(2)
    return df

#Llamamos a la función para modificar la lista
df_data=object_to_float(df_data, list_object_float)

4.13 Estandarizamos columnas object a integer.

In [269]:
#Cambiamos las columnas de la lista list_object_integer
df_data[['Employee_number', 'Performance_rating', 'Total_working_years', 'Work_life_balance', 'Years_in_current_role']] = df_data[['Employee_number', 'Performance_rating', 'Total_working_years', 'Work_life_balance', 'Years_in_current_role']] .replace(',0', '', regex=True).astype(float).astype(pd.Int64Dtype()) #esta parte permite convertir a integer y que se mantengan los nulos existentes


4.14 Unificamos todos los textos en .capitalize().

In [270]:
cat_cols=df_data.select_dtypes(include=['O'])
#Función para pasar todas las celdas a capitalize()
def capitalize_string(df, columns):
    for i in columns:
        df[i]=df[i].str.capitalize()
    return df

#Llamamos a la función para modificar la lista
df_data=capitalize_string(df_data, cat_cols)

### 5. Gestión de nulos y eliminación de columnas prescindibles

5.1 Código para calcular nulos y sus %.

In [271]:
#Separamos las columnas por categóricas y numéricas
cat_cols=df_data.select_dtypes(include=['O'])
num_cols=df_data.select_dtypes(include=['number'])

#Calculamos la cantidad y % de nulos de las columnas categóricas
cat_nulos_df=pd.DataFrame({'Column': cat_cols.columns, #trae los nombres de las columnas
                            'Nulos': cat_cols.isnull().sum(), #trae la suma de los nulos
                            '% Nulos': (cat_cols.isnull().mean() * 100).round(2)}) #trae el % de nulos
cat_nulos_df=cat_nulos_df[cat_nulos_df['Nulos'] > 0] #filtra para aquellos >0

#Calculamos la cantidad y % de nulos de las columnas numéricas
num_nulos_df=pd.DataFrame({'Column': num_cols.columns, #trae los nombres de las columnas
                            'Nulos': num_cols.isnull().sum(), #trae la suma de los nulos
                            '% Nulos': (num_cols.isnull().mean() * 100).round(2)}) #trae el % de nulos

#Creamos variables en las que guardamos las medias y medianas de cada columna numérica
num_mean=num_cols.mean()
num_median=num_cols.median()
#Añadimos las medias y medianas al df de columnas numéricas
num_nulos_df['Mean']=num_nulos_df['Column'].map(num_mean)
num_nulos_df['Median']=num_nulos_df['Column'].map(num_median)

num_nulos_df=num_nulos_df[num_nulos_df['Nulos'] > 0] #filtra para aquellos >0

In [272]:
cat_nulos_df

Column  Nulos  % Nulos
Business_travel    Business_travel    772    47.83
Department              Department   1312    81.29
Education_field    Education_field    745    46.16
Marital_status      Marital_status    651    40.33
Over_18                    Over_18    901    55.82
Over_time                Over_time    676    41.88
Role_departament  Role_departament   1312    81.29
Remote_work            Remote_work    345    21.38

5.2 Toma de decisiones para la gestión de nulos.

Observando los nulos y sus %, decidimos eliminar las siguientes columnas:
- Department 81.29% nulos.
- Role_departament 81.29% nulos.

In [273]:
df_data.drop(['Department', 'Role_departament'], axis=1, inplace=True)

Observando los altos % de nulos en el resto de columnas categóricas, decidimos sustituir los valores NaN por 'Unknown' para no alterar la veracidad de los datos.

In [274]:
#Reemplazamos los nulos con 'unknown'
columns_unknown=['Business_travel', 'Education_field', 'Marital_status', 'Over_18', 'Over_time', 'Remote_work']
for i in columns_unknown:
    df_data[i]=df_data[i].fillna('Unknown')

In [275]:
num_nulos_df

Column  Nulos  % Nulos         Mean  \
Daily_rate                          Daily_rate    124     7.68   802.085235   
Employee_number                Employee_number    431    26.70  1000.289941   
Hourly_rate                        Hourly_rate     84     5.20    66.028105   
Monthly_income                  Monthly_income    843    52.23  6477.686122   
Performance_rating          Performance_rating    195    12.08     3.150810   
Standard_hours                  Standard_hours   1195    74.04    80.000000   
Total_working_years        Total_working_years    526    32.59    11.318934   
Work_life_balance            Work_life_balance    108     6.69     2.759628   
Years_in_current_role    Years_in_current_role   1580    97.89     4.029412   
Same_as_monthly_income  Same_as_monthly_income    843    52.23  6477.686122   
Number_children                Number_children   1614   100.00          NaN   

                        Median  
Daily_rate               798.0  
Employee_number          972.0  
Hourly_rate               66.0  
Monthly_income          4859.0  
Performance_rating         3.0  
Standard_hours            80.0  
Total_working_years       10.0  
Work_life_balance          3.0  
Years_in_current_role      2.0  
Same_as_monthly_income  4859.0  
Number_children            NaN

Observando los nulos y sus %, decidimos eliminar las siguientes columnas:

- Monthly_income 52.23% nulos.
- Standard_hours 74.04 % nulos y solamente tiene un valor único (80).
- Years_in_current_role 97.89% nulos.
- Same_as_monthly_income es idéntica a monthly_income
- Number_children 100% nulos
- Salary. No tiene nulos pero todos sus valores son idénticos (no es realista).

In [276]:
df_data.drop(['Monthly_income', 'Standard_hours', 'Years_in_current_role', 'Same_as_monthly_income', 'Number_children', 'Salary'], axis=1, inplace=True)

Columna Employee_number:

Tiene un 26.70% nulos. Para que la empresa pueda seguir utilizando este identificador a nivel interno, decidimos mantener esta columna pero, puesto que utilizaremos un ID autoincremental para nuestro análisis, convertimos la columna a object y sustituimos los NaN con 'unknown'.

In [277]:
#Convertimos la columna a object
df_data['Employee_number']=df_data['Employee_number'].astype('object')

#Sustituimos los nullos por 'unkown'.
df_data['Employee_number']=df_data['Employee_number'].fillna('Unknown')

Para las columnas con un % de nulos inferior al 10%, aplicaremos la sustitución de esos nulos por la media o la mediana. Comparando éstas, y al ser tan parecidas entre sí, optamos por sustituir en base a la media.

In [278]:
#Reemplazamos nulos con la media
columns_mean=['Daily_rate', 'Hourly_rate', 'Work_life_balance']
for i in columns_mean:
    imputer=SimpleImputer(strategy='mean')
    df_data[i]=imputer.fit_transform(df_data[[i]])
    df_data[i]=round(df_data[i], 2)

Para las columnas con un % de nulos superior al 10% aplicamos un IterativeImputer (random_state=42) para sustituirlos con una estimación lo más precisa posible.

In [279]:
#Reemplazamos nulos con IterativeImputer
columns_iterative_imputer=['Performance_rating', 'Total_working_years']
for i in columns_iterative_imputer:
    imputer=IterativeImputer(max_iter=20, random_state=42)
    df_data[i]=imputer.fit_transform(df_data[[i]])
    df_data[i]=round(df_data[i], 2)

In [280]:
#Comprobamos que hemos eliminado todos los nulos
df_data.isnull().sum()

Age                           0
Attrition                     0
Business_travel               0
Daily_rate                    0
Distance_from_home            0
Education                     0
Education_field               0
Employee_count                0
Employee_number               0
Environment_satisfaction      0
Gender                        0
Hourly_rate                   0
Job_involvement               0
Job_level                     0
Job_role                      0
Job_satisfaction              0
Marital_status                0
Monthly_rate                  0
Num_companies_worked          0
Over_18                       0
Over_time                     0
Percent_salary_hike           0
Performance_rating            0
Relationship_satisfaction     0
Stock_option_level            0
Total_working_years           0
Training_times_last_year      0
Work_life_balance             0
Years_at_company              0
Years_since_last_promotion    0
Years_with_curr_manager       0
Date_bir

In [281]:
#Visualizamos la info final.
df_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1614 entries, 0 to 1613
Data columns (total 33 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Age                         1614 non-null   int64  
 1   Attrition                   1614 non-null   object 
 2   Business_travel             1614 non-null   object 
 3   Daily_rate                  1614 non-null   float64
 4   Distance_from_home          1614 non-null   float64
 5   Education                   1614 non-null   int64  
 6   Education_field             1614 non-null   object 
 7   Employee_count              1614 non-null   float64
 8   Employee_number             1614 non-null   object 
 9   Environment_satisfaction    1614 non-null   int64  
 10  Gender                      1614 non-null   object 
 11  Hourly_rate                 1614 non-null   float64
 12  Job_involvement             1614 non-null   int64  
 13  Job_level                   1614 

### 6. Posibles mejoras a añadir en el csv:
- Categorizamos algunas columnas para una mejor comprensión
```python
#Función para clasificar la distancia
def classify_distance(km):
    if km < 10:
        return 'near'
    elif km < 20:
        return 'medium'
    else:
        return 'far'

df_data['Distance_classification']=df_data['Distance_from_home'].apply(classify_distance)
```

### 7. Exportamos csv.

In [282]:
df_data.to_csv('hr_data_final.csv')